keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+can cause, or contribute to, request timeouts being exceeded.
+
+Webhooks must either be configured as `FailPolicy.Fail` or `FailPolicy.Ignore`. If
+`FailPolicy.Ignore` is used, there is potential for requests skip the webhook and
+be admitted. If `FailPolicy.Fail` is used, a webhook outage can result in a
+localized or widespread Kubernetes control plane outage depending on which
+objects the webhook is configured to intercept.
+
+### Goals
+
+- Make CRDs more self-contained and declarative
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+- Support for validation formats, immutability or unions. These are all valuable improvements
+  but can be addressed orthogonally in separate KEPs.
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. For example, if a validation check requires making a network
+  request to some other system, it will still need to be implemented in a webhook.
+- Support all validation done on native Kubernetes types. For example, CRD structural schemas has
+  complex validation rules that we CEL cannot support due to the lask of support for arbitrarily
+  deeply nested terms (CEL cannot support recursive data types).
+- Change how Kubernetes built-in types are validated, defaulted and converted.
+
+## Proposal
+
+An inline expression language like [Common Expression Language (CEL)](https://github.com/google/cel-go) would be an
+excellent supplement to the current validation mechanism because it is sufficiently expressive to
+satisfy a large set of remaining uses cases that none of the above can solve.
+For example, cross-field validation use cases can only be solved using
+expressions or webhooks.
+
+`x-kubernetes-validator` extension will be added to CRD structural schemas to allow CEL validation expressions.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          x-kubernetes-validator: 
+            - rule: "minReplicas <= maxReplicas"
+              message: "minReplicas cannot be larger than maxReplicas"
+          type: object
+          properties:
+            minReplicas:
+              type: integer
+            maxReplicas:
+              type: integer
+```
+
+- Each validator may have multiple validation rules.
+
+- Each validation rule has an optional 'message' field for the error message that
+will be surfaced when the validation rule evaluates to false.
+
+- The validator will be scoped to the location of the `x-kubernetes-validator`
+extension in the schema. In the above example, the validator is scoped to the
+'spec' field.
+
+- For OpenAPIv3 object types, the expression will have direct access to all the
+fields of the object the validator is scoped to.
+  
+- For OpenAPIv3 scalar types (integer, string & boolean), the expression will have
+access to the scalar data element the validator is scoped to.
```

Comment:
How about `field`?

I've gone back and forth on this. The main alternatives I've considered are:

1. Reserve a symbol (`_`, or `this` or something)
2. Let the author of the validation rule pick (one of the fields for declaring the validation rule sets the identifier used to reference the data element)
3. Use the field name (e.g. if the validation rule is on a field named `labelSelector`, then put `labelSelector` in scope)

I honestly done feel super strong about this, and it seems quite subjective.  I wanted to avoid `this` due to the semantics/baggage from Javascript and Java. `_` just seems a bit to obtuse.

I'll pencil this into the KEP but am interested in hearing ideas from others.

Reply:
3 seems reasonable, makes for expressions that read naturally, and remain correct if the rule is moved up a level to the parent object in order to compare to some peer field

Reply:
👍 #3 works out well. Updating the KEP to match.